API: Restore implicit converter registration

TomAugspurger opened the pull request:
Closes #18301 

TODO:

- [ ] find best home for the import
- [ ] update deprecation message

TomAugspurger committed:
API: Restore implicit converter registration

TomAugspurger commented:
I added the "special" section in the whatsnew for this and the changes we're making to NaN/empty sum.

TomAugspurger committed:
Remove matplotlib from blacklist

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,45 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+
+This caused problems for some users, so we're temporarily reverting that change;
+pandas will again register the converters on import. Using the converters
+without explicitly registering the formatters will cause a ``FutureWarning``:
```
I would not use 'the formatters' as it actually points to the same as 'converters' but with another word, just making it confusing. I would either just leave it out or use 'them' to avoid repeating converters

TomAugspurger replied:
Meant to use formatters everywhere. Fixing.

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,45 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+           >>> from pandas.tseries import converter
+           >>> converter.register()
+
+As the error message says, you'll need to register the converts if you intend to
```
converts -> converters

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,45 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+
+As the error message says, you'll need to register the converts if you intend to
+use them with matplotlib plotting functions. Pandas plotting functions, such as
+``Series.plot``, will register them for you. (:issue:`18301`)
```
maybe a bit more explicit like: "When using pandas plotting functions, such as Series.plot, the converters will be registered for you and you do not need to do this explicitly"

jorisvandenbossche commented on pandas/plotting/_core.py:
```diff
@@ -2352,7 +2354,6 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
     >>> grouped = df.unstack(level='lvl1').groupby(level=0, axis=1)
     >>> boxplot_frame_groupby(grouped, subplots=False)
     """
-    _setup()
```
this one doesn't need to be replaced with _converter.WARN = False ?

jorisvandenbossche commented on pandas/tests/plotting/test_warns.py:
```diff
@@ -0,0 +1,35 @@
+plt = pytest.importorskip('matplotlib.pyplot')
+
+
+def test_warns():
```
Is this duplicate with the other test file?

TomAugspurger replied:
Yes, forgot to remove it from my first commit.

jorisvandenbossche reviewed (commented):
looks good to me!

TomAugspurger commented:
What are people's thoughts on moving the `register` import to `pandas.plotting` instead of `pandas.tseries`? I'm not sure I see the benefit, unless we're planning to remove `pandas.tseries` in the future.

TomAugspurger committed:
fixup! Remove matplotlib from blacklist

jorisvandenbossche commented:
Should we also add some language on how to avoid this warning *without* registering the converters by default? So if you in the future don't want to manually register them, but rely on the matplotlib ones (eg when working with datetime.datetime) ?

TomAugspurger commented:
I guess a pandas option?

```
pd.options.register_formatters : bool
```

And we check that before ever calling register?

TomAugspurger commented:
Though, future-proofing a bit, should that be spread across several options, one per type? We currently register for

```
    units.registry[lib.Timestamp] = DatetimeConverter()
    units.registry[Period] = PeriodConverter()
    units.registry[pydt.datetime] = DatetimeConverter()
    units.registry[pydt.date] = DatetimeConverter()
    units.registry[pydt.time] = TimeConverter()
    units.registry[np.datetime64] = DatetimeConverter()
```

In the ideal future, we would I think just do Timestamp and Period, and matplotlib would have datetime64, I don't know who wants to own python dates / times / datetimes.

cc @tacaswell if any of this interests you.

TomAugspurger committed:
Add option for toggling formatters

TomAugspurger commented:
Added a rough implementation of the option in 16c33f1.


```python
In [1]: import pandas as pd

In [2]: import matplotlib.pyplot as plt

In [3]: fig, ax = plt.subplots()

In [4]: s = pd.Series(range(12), index=pd.date_range('2017', periods=12))

In [5]: pd.options.plotting.matplotlib.register_formatters = False

In [6]: ax.plot(s)
Out[6]: [<matplotlib.lines.Line2D at 0x1131ce588>]
```

TomAugspurger committed:
Remove move

jreback commented:
> What are people's thoughts on moving the register import to pandas.plotting instead of pandas.tseries? I'm not sure I see the benefit, unless we're planning to remove pandas.tseries in the future.

if its not too much trouble. I would like to at least deprecate ``pandas.tseries``

TomAugspurger committed:
Handle no matplotlib

tacaswell commented:
Thanks :heart: .

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,50 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+
+As the error message says, you'll need to register the converters if you intend
+to use them with matplotlib plotting functions. Pandas plotting functions, such
+as ``Series.plot``, will register them for you calling ``converter.register()``
```
"for you calling" -> "for you, and calling"

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,50 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+as ``Series.plot``, will register them for you calling ``converter.register()``
+first is not necessary.
+
+Finally, control the formatters, we've added a new option:
```
"to" control

jorisvandenbossche commented:
Some other things related to the option:

- It seems you kept the terminology of 'converters' (and not 'formatters') throughout the whatsnew and code? (which is I think the terminology of matplotlib, as the converter calls a formatter). But the option is called 'register_formatters'. So we should choose here.

- When you set the `pd.options.plotting.matplotlib.register_formatters = False`, they will also not get registered when using pandas' plot methods ? I don't think that is the case, but shouldn't it be?

- The "unregistering" (setting `pd.options.plotting.matplotlib.register_formatters = False`) will also remove the converters for datetime.datetime, which are actually the default ones. Ideally they are set back as well. @tacaswell Is there a way to restore the default units? (like with rcParams). Of course we can also save them when we overwrite them, to set it back properly.

- I would make the option maybe a bit shorter, like `pd.options.plotting.register_matplotlib_formatters` (not shorter, but a '.' less so quicker to tab-complete) ?

jorisvandenbossche commented on pandas/tests/plotting/test_converter.py:
```diff
@@ -15,6 +16,75 @@ def test_timtetonum_accepts_unicode():
+        converter.register()
+        with ctx:
+            with tm.assert_produces_warning(None) as w:
+                ax.plot(s.index, s.values)
```
If you did set the option to False, it should not warn?

jorisvandenbossche replied:
Sorry, forget, that is what is being tested :-) (and they have been registered manually afterwards)

Should we test what happens if you do not register manually after you set the option?

TomAugspurger replied:
You mean set the option to True or False? I think my latest commit does that.

After setting it to False, I think we should not warn even if they don't register manually (which is what my tests checks hopefully)

TomAugspurger commented:
> It seems you kept the terminology of 'converters' (and not 'formatters') 

Switched to formatters everywhere.

> they will also not get registered when using pandas' plot methods ?

Correct. To be more specific, they're *always* registered when pandas is imported. The option toggles whether or not they're in the `matplotlib.units.registry` dict. `True` adds them, `False` removes them. That means that doing

```python
In [8]: pd.options.plotting.matplotlib.register_formatters = False

In [9]: s.plot()
```

Will fall back to whatever matplotlib does with datetimes / periods (potentially an exception, e.g. for periods).

> will also remove the converters for datetime.datetime, which are actually the default ones

Hmm, not sure about that. My registry is empty upon import:

```python
In [1]: import matplotlib.units

In [2]: matplotlib.units.registry
Out[2]: {}
```

Though maybe I'm misunderstanding. The intent is return the registry to the state before we mucked with it.

> I would make the option maybe a bit shorter

Yeah, thoughts on `pd.options.plotting.mpl.converters`? I want the `.mpl` / matplotlib namespace for if / when we allow different engines for `.plot`.

TomAugspurger committed:
Cleanup

TomAugspurger committed:
Test no register

tacaswell commented:
The unit registry gets filled on import.  If you import `matplotlib.dates` they will be added.  These imports are done in `pyplot` so _most_ users have this import done.

TomAugspurger commented:
Ah thanks.

```python
In [1]: import matplotlib.units

In [2]: import matplotlib.pyplot as plt

In [3]: matplotlib.units.registry
Out[3]:
{str: <matplotlib.category.StrCategoryConverter at 0x10beb7f60>,
 numpy.str_: <matplotlib.category.StrCategoryConverter at 0x10beb7f28>,
 bytes: <matplotlib.category.StrCategoryConverter at 0x10beb7ef0>,
 numpy.bytes_: <matplotlib.category.StrCategoryConverter at 0x10beb7f98>,
 datetime.date: <matplotlib.dates.DateConverter at 0x10bf2c208>,
 datetime.datetime: <matplotlib.dates.DateConverter at 0x10bf2c240>}

In [4]: import pandas as pd

In [5]: matplotlib.units.registry
Out[5]:
{str: <matplotlib.category.StrCategoryConverter at 0x10beb7f60>,
 numpy.str_: <matplotlib.category.StrCategoryConverter at 0x10beb7f28>,
 bytes: <matplotlib.category.StrCategoryConverter at 0x10beb7ef0>,
 numpy.bytes_: <matplotlib.category.StrCategoryConverter at 0x10beb7f98>,
 datetime.date: <pandas.plotting._converter.DatetimeConverter at 0x10bf2c240>,
 datetime.datetime: <pandas.plotting._converter.DatetimeConverter at 0x10eaa3828>,
 pandas._libs.tslib.Timestamp: <pandas.plotting._converter.DatetimeConverter at 0x10eaa3780>,
 pandas._libs.period.Period: <pandas.plotting._converter.PeriodConverter at 0x10eaa37b8>,
 datetime.time: <pandas.plotting._converter.TimeConverter at 0x10bf2c208>,
 numpy.datetime64: <pandas.plotting._converter.DatetimeConverter at 0x10eaac0f0>}

In [6]: pd.options.plotting.matplotlib.register_formatters = False

In [7]: matplotlib.units.registry
Out[7]:
{str: <matplotlib.category.StrCategoryConverter at 0x10beb7f60>,
 numpy.str_: <matplotlib.category.StrCategoryConverter at 0x10beb7f28>,
 bytes: <matplotlib.category.StrCategoryConverter at 0x10beb7ef0>,
 numpy.bytes_: <matplotlib.category.StrCategoryConverter at 0x10beb7f98>}
```

So not great... We could of course maintain our own global state when we overwrite them... Though we'll have to make sure `matplotlib.pyplot` is imported first. Let me give that a shot.

TomAugspurger committed:
Restore original state

TomAugspurger committed:
Added deregister

TomAugspurger committed:
Doc, naming

TomAugspurger committed:
Naming

TomAugspurger commented:
Just pushed a few changes, if you have a chance to review again.

- Moved recommended import to `pandas.plotting.register_formatters`, `pandas.plotting.deregister_formatters`
- Added a `deregister_formatters` to undo what pandas does (needed it anyway for the callback, might as well make it public)
- Updated the release note / error message to always use the term "converter"


TODO:

- [ ] add depr warning in tseries
- [ ] *Another* PR to matplotlib updating their FAQ (sorry)

Should be able to do those later today.

TomAugspurger committed:
Added deprecation

pep8speaks commented:
Hello @TomAugspurger! Thanks for updating the PR.













 - In the file [`pandas/tseries/converter.py`](https://github.com/pandas-dev/pandas/blob/5ee385e9af84bb75ecd1de3e03759db7a1ddb323/pandas/tseries/converter.py), following are the PEP8 issues :

> [Line 21:1](https://github.com/pandas-dev/pandas/blob/5ee385e9af84bb75ecd1de3e03759db7a1ddb323/pandas/tseries/converter.py#L21): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

TomAugspurger committed:
PEP8

TomAugspurger committed:
Fix typos

jorisvandenbossche commented:
@TomAugspurger Thanks a lot for the updates!

> Moved recommended import to pandas.plotting.register_formatters, pandas.plotting.deregister_formatters

+1. I was only wondering if we should make it more explicit it is for matplotlib (like `register_matplotlib_converters`)? (although I know I previously wondered if it could be shorter ... :-), but for case matplotlib is not the only supported plotting engine in the future)

> Yeah, thoughts on pd.options.plotting.mpl.converters? I want the .mpl / matplotlib namespace for if / when we allow different engines for .plot.

OK, yes, having 'matplotlib' in there makes sense. Although I prefer the full 'matplotlib' instead of `mpl`. And I think I find 'register_converters = True/False' clearer than `converters = True/False`. Although then we are back at the original `pd.options.plotting.matplotlib.register_converters` :-)

jorisvandenbossche commented:
Something else: I was wondering if it actually *would* be possible to do the registration/deregistration in a context manager in our plotting machinery. The reason we said this would be impossible is that you after a `df.plot` call you want to use matplotlib's machinery to tune the figure (eg `ax.set_xlim(..)`), so this still needs to know about the converters. 
But, it seems that the `ax` object actually stores the used converters for plotting. So maybe that would be enough?

```
In [15]: s = pd.Series(range(10), pd.date_range("2012-01-01", periods=10))

In [16]: ax = s.plot()

In [17]: ax.xaxis.converter
Out[17]: <pandas.plotting._converter.PeriodConverter at 0x7f3c112f9ac8>
```

jorisvandenbossche commented:
As a small example:

```
In [1]: s = pd.Series(range(10), pd.date_range("2012-01-01", periods=10))

In [3]: import matplotlib.pyplot as plt

In [4]: fig, ax = plt.subplots()

In [5]: s.plot(ax=ax)
Out[5]: <matplotlib.axes._subplots.AxesSubplot at 0x7fd4580d4fd0>

In [6]: pd.plotting.deregister_converters()

In [7]: ax.xaxis.converter
Out[7]: <pandas.plotting._converter.PeriodConverter at 0x7fd45d4538d0>

In [8]: ax.set_xlim("2012-01-01", "2012-01-04")    # <---- still works although converters are no longer registered
Out[8]: (15340, 15343)

In [10]: import matplotlib

In [11]: matplotlib.units.registry
Out[11]: 
{bytes: <matplotlib.category.StrCategoryConverter at 0x7fd472ac6278>,
 datetime.date: <matplotlib.dates.DateConverter at 0x7fd45d1fd6d8>,
 numpy.bytes_: <matplotlib.category.StrCategoryConverter at 0x7fd472ac6320>,
 str: <matplotlib.category.StrCategoryConverter at 0x7fd472ac62e8>,
 numpy.str_: <matplotlib.category.StrCategoryConverter at 0x7fd472ac62b0>,
 datetime.datetime: <matplotlib.dates.DateConverter at 0x7fd45d1fd710>}
```

@tacaswell any idea there would be caveats to this approach? Are there customizations that you can do on plots that will typically not work like that? (that would not be using the stored `converter` on the axis method)

TomAugspurger commented:
Oh, that's neat.

But, do we want to do that for this PR or later on? And what would that look like? How about the option takes three values:

- True  (register right away)
- False (don't register / unregister)
- auto (Use a context manager in our plots)

The current default will be True which is compatible with <0.21.0. In the future it'll be 'auto'.

Though I'm not sure I'll have time to implement the 'auto' behavior today, and I'll be traveling the rest of the week.

TomAugspurger committed:
Rename it all

jorisvandenbossche commented:
> But, do we want to do that for this PR or later on?

I don't think it needs to be in this PR (as this already introduces a lot of useful architecture), but if we want to do it, I think we should try to do it for 0.21.1, as it will again be a change in behaviour (the change from True (once plotted once) to auto).

>  And what would that look like? How about the option takes three values:

I think those three options are good: either registerer it matplotlib-wide, either never registerer it, or only register it temporarily for plots that are made with pandas.

But we will probably still need to look at a couple of edge cases if it actually works.

I would say let's try to merge this PR for now, and leave the 'auto' behaviour for a follow-up. I could also try to look at it this week.

TomAugspurger committed:
Missed one

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into depr-converter

jorisvandenbossche commented:
Another thing I was thinking: for users that were just relying on the converter by importing pandas (but for the rest plotting with pure matplotlib), deprecating it now can be a bit annoying, as the proper support in matplotlib will only come in the next release. 
So in practice that means that they will have to add temporarily `pd.plotting.register_matplotlib_converters()` to their code for a couple of months, instead of directly being able to switch to the new matplotlib support.

So that said, another option is to just undo the change now for 0.21.1 and keep the actual deprecation for 0.22.0 (assuming that will only be released when matplotlib 2.2.0 will also more or less be released). 
@tacaswell Is there a timeline for matplotlib 2.2.0? Or could the datetime64 converter be sneaked into 2.1.1? (considering it as kind of a 'regression fix' to deal with the changes in pandas ..)

(just a thought, not sure yet if I find postponing actual deprecation a good idea myself)

jorisvandenbossche commented:
@TomAugspurger What do you think of the idea above to completely revert this for 0.21.1, and target this PR for 0.22.0, in the idea that by then matplotlib has the datetime64 support, and many users won't need to do the manual registration?

jreback commented:
had a talk with @tacaswell about this. are we sure we want this behavior?

jorisvandenbossche commented:
What do you mean with "this behavior"?

Ideally (for pandas), matplotlib would already include the datetime64 converters in 2.1.1 bugfix release (the merged PR there was now targetted for matplotlib 2.2). But I assume they are not favoring that since it not a bugfix from the matplotlib point of view (if you see @tacaswell you could bring that up)

jreback commented:
> Ideally (for pandas), matplotlib would already include the datetime64 converters in 2.1.1 bugfix release (the merged PR there was now targetted for matplotlib 2.2). But I assume they are not favoring that since it not a bugfix from the matplotlib point of view (if you see @tacaswell you could bring that up)

yes, we changed and let's move on. mpl will catch up with the fix. reverting just makes this even more confusing.

jorisvandenbossche commented:
This PR *is* already reverting it (and we agree on doing this I thought), the question is more to already raise a deprecation warning or not?

TomAugspurger commented:
Just thinking out loud here to make sure I understand:

1. We're going to register the converters again on `import pandas` either way
2. We're uncertain on whether to include the deprecation warning for using an implicitly registered converter for datetime data.

The uncertainty with 2 comes from us *maybe* deprecating `converter.register`. We don't want to say "you're required to register this" and then remove that shortly after, right?

I think pandas will always have to have *some* converters, if just for things like `PeriodIndex`, which matplotlib has no knowledge of. So I think my vote is for including the deprecation warning.

TomAugspurger commented:
One more issue...  If someone with mpl2.2+ (with their datetime converters), we should *not* warn when they plot datetimes and have imported pandas, but have not registered the converters. I think...

jorisvandenbossche commented:
> The uncertainty with 2 comes from us maybe deprecating converter.register. We don't want to say "you're required to register this" and then remove that shortly after, right?

No, my reason to *delay* deprecation is that for the majority of the users (that do like `plt.plot(s.index, s.values` with DatetimeIndex) the matplotlib converters will be sufficient (I think). 
For this part, if we release 0.21.1 with deprecation warning, we ask them to manually register the converters. But once matplotlib 2.2 comes out somewhere next year, they don't actually need this manually registering of the pandas converters anymore.  
If we wait with the deprecation until pandas 0.22.0 (assuming that matplotlib 2.2 is released by then), they will never need to bother with registering the converters

> I think pandas will always have to have some converters, if just for things like PeriodIndex, which matplotlib has no knowledge of. So I think my vote is for including the deprecation warning.

yes, and for those we will need to do that. But I think the majority of the usecase is DatetimeIndex and not PeriodIndex.

jorisvandenbossche commented:
> One more issue... If someone with mpl2.2+ (with their datetime converters), we should not warn when they plot datetimes and have imported pandas, but have not registered the converters. I think...

Yes, and exactly for that reason I would wait until matplotlib 2.2+ is available

TomAugspurger commented:
> they will never need to bother with registering the converters

OK, that's the bit I was missing. I'll make a new PR with just the restoration, and leave this one till after MPL 2.2 (January).

TomAugspurger committed:
Check version

jorisvandenbossche commented:
It might be we can already merge this, but just set warn to False and the default value of `pd.options.plotting.matplotlib.register_converters` to True ?  (but not sure that will completely work?)

The functionality to turn them off completely can already be useful?

jreback commented:
> pd.options.plotting.matplotlib.register_converters to True

I have no problem with providing this optinon, but why would you make this True by default? it is conceptually false now.

jorisvandenbossche commented:
> I have no problem with providing this optinon, but why would you make this True by default? it is conceptually false now.

It has always been conceptually `True` (up to 0.21.0), so if we revert, then the default would be True. And from 0.21.0 you could say that it is conceptually False, but that is also not really correct, as it is False on import but True once you do plotting with pandas. In this PR, the `False` option will also not register them when plotting datetimes (so that will raise an error with current matplotlib)

TomAugspurger commented:
Another issue for down the road https://github.com/matplotlib/matplotlib/pull/9779 doesn't handle `DatetimeIndexes`:

<img width="970" alt="screen shot 2017-11-29 at 10 09 36 am" src="https://user-images.githubusercontent.com/1312546/33381973-7ee1ee82-d4ed-11e7-837a-18da9be6a2ea.png">

Though hopefully this can be handled on the matplotlib side by getting the `values`.

jorisvandenbossche commented:
Yes, I think that should be easily solvable in matplotlib. I was still planning to run all the different reported examples in pandas issues related to this with matplotlib master to see if it would solve those cases.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into depr-converter

TomAugspurger commented:
For MPL 2.2+, do we want to overwrite their converters with our own? I think no, but I'll have to check the differences.

jorisvandenbossche commented:
To summarize, I think in the long term we want those options:

- False: never register our converters (also not when pandas plotting is used)
- True: register our converters, so also available for pure matplotlib plots
- 'auto': only use our converters for plots made by our `.plot(..)` method. Here I think we still want to overwrite matplotlib's converters (but that's maybe up for discussion)

Before 0.21.0, 'True' was the default, and in the long term we want 'auto' to become the default. 

But as a transition period, I think we want a hybrid of True/'auto', which would be 'auto' + deprecation warning for implicitly used converters (i.e. register our converters but with warn=True). And indeed there I would only do this warning for those types that are in the meantime not supported by matplotlib itself, and thus not overwrite the ones matplotlib in the meantime added. 
Not overwriting those can still mean some breaking differences for users, but much less as is the case now (the underlying plotted float data are different, but in principle a normal user never cares about that + the formatting of the axes will be a bit different)

But, my proposal it to now do a complete revert for 0.21.1, and start the transition period only in 0.22.0 (or the first release when matplotlib 2.2 is out).

TomAugspurger committed:
No warnings by default

TomAugspurger committed:
Update release notes

TomAugspurger commented:
Sorry for the delay on this. I'm finally done with traveling and conferences for a bit.

This PR now

- Adds the options for easily adding / removing our converters
- Moved `pandas.tseries.register` to `pandas.plotting.register_matplotlib_converters`
- Adds the scaffolding for warning in the next version about implicitly registered formatters, but doesn't actually use this yet, as the default is to not warn

So, after MPL 2.2, we should just need to change the default in `register` to be True, in which case we'll warn on matplotlib plotting methods.

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,39 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+for regular ``matplotlib.pyplot`` plotting methods, so we're temporarily
+reverting that change; pandas will again register the converters on import.
+Pandas plotting functions, such as ``Series.plot``, will continue to register
+them for you.
```
This sentence now feels a bit out of place, as the previous one already said that they are registered again on import.

jorisvandenbossche commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -7,6 +7,39 @@ This is a minor release from 0.21.1 and includes a number of deprecations, new
+.. code-block:: python
+
+   >>> from pandas.plotting import register_matplotlib_converters
+   >>> register_matplotlib_converters()
```
I am not sure if I would show that here. Because people who glance over it might think they need to add that to their code, which is not the case.

TomAugspurger committed:
Test fixup

- actually switch the default to not warn
- We do overwrite matplotlib's formatters

jorisvandenbossche commented on pandas/tseries/converter.py:
```diff
@@ -9,3 +10,11 @@
+def register():
+    from pandas.plotting._converter import register as register_
+    msg = ("'pandas.tseries.converter.register' has been moved and renamed to "
+           "'pandas.plotting.register_converters'. ")
```
register_converters -> register_matplotlib_converters()

jorisvandenbossche reviewed (commented):
This is looking good to me, thanks a lot!

Some minor comments. 
Should we add a test that currenlty on import they are already registered?

jreback commented:
@tacaswell can you have a look

TomAugspurger commented:
>  Should we add a test that currenlty on import they are already registered?

I thought about that, but there are test-order issues with  that. Do you know a clean way of running a single test in it's own process? I could use subprocess, but that seems fragile.

TomAugspurger committed:
Doc update

TomAugspurger committed:
Fix deprecation message

jorisvandenbossche commented:
> I thought about that, but there are test-order issues with that. Do you know a clean way of running a single test in it's own process? I could use subprocess, but that seems fragile.

Not directly, googling said pytest does not have a way to specify this, but found this: https://stackoverflow.com/questions/45462374/mark-test-to-be-run-in-independent-process (not sure how robust it is)

TomAugspurger committed:
Test added by default

TomAugspurger commented:
Added https://github.com/pandas-dev/pandas/pull/18307/commits/ee7b457f01c4489e351c4567340da86e22ab16a6 with a test using subprocess. I worry that's fragile to PATH issues, but let's see.

tacaswell reviewed (approved):
I am happy with this long term plan.

jorisvandenbossche commented:
@TomAugspurger Let's merge this?

jorisvandenbossche commented:
I opened https://github.com/pandas-dev/pandas/issues/18720 as a follow-up issue about the 'auto' behaviour (only modifying matplotlibs units registry in pandas' plotting context)
